exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +325,255 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		return err
+	}
+
+	for i := range machineList.Items {
+		machine := &machineList.Items[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			// Until the InfraMachinePool populates its selector kind and selector labels, the MachinePool controller can't tell that it supports MachinePool Machines
+			// and will create fallback machines to begin with. When the selector fields gets populated, we want to clean up those fallback machines.
+			log.V(2).Info("Machine has an empty infraRef, this a fallback machine, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+
+			return r.Client.Delete(ctx, machine)
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(2).Info("Machine is not orphaned, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, infraMachines []unstructured.Unstructured) error {
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	setOfInfraRefNames := map[string]struct{}{}
+	for _, machine := range machineList.Items {
+		infraRef := machine.Spec.InfrastructureRef
+		setOfInfraRefNames[infraRef.Name] = struct{}{}
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		if _, infraRefExists := setOfInfraRefNames[infraMachine.GetName()]; infraRefExists {
+			continue
+		}
+
+		machine := getNewMachine(mp, infraMachine, false)
+		if err := r.Client.Create(ctx, machine); err != nil {
+			return errors.Wrapf(err, "failed to create new Machine for infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace())
+		}
+
+		// Set the owner reference on the infraMachine to the Machine since the infraMachine is created and owned by the infraMachinePool.
+		infraMachine.SetOwnerReferences([]metav1.OwnerReference{
```

Comment:
I have some questions about this. I'm working on a CAPA infrastructure provider implementation for this proposal, and I noticed `metav1.NewControllerRef` returns a `OwnerReference` that has `blockOwnerDeletion` set to `true`. For example, in my AWS Machines, I see:
```
kind: AWSMachine
metadata:
...
    ownerReferences:
    - apiVersion: cluster.x-k8s.io/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: Machine
      name: worker-az1-tb8v5
      uid: 23e24426-4dc1-412d-8d99-96cb693a6ba5
```

Should `blockOwnerDeletion` be set to false on this owner reference? I ask because currently when the AWS ASG scales down and terminates an ec2 node, and then the CAPA AWSMachinePool reconciler cleans up the dangling AWSMachine, the owner Machine is not cleaned up because the owner deletion is disabled.

Or do I misunderstand how this is supposed to work?

Reply:
> Kubernetes also adds finalizers to an owner resource when you use either [foreground or orphan cascading deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#cascading-deletion). In foreground deletion, it adds the foreground finalizer so that the controller must delete dependent resources that also have ownerReferences.blockOwnerDeletion=true before it deletes the owner. If you specify an orphan deletion policy, Kubernetes adds the orphan finalizer so that the controller ignores dependent resources after it deletes the owner object.

Based on this quote from [this page](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/), my understanding is that `blockOwnerDeletion` means that the "children" of a CRD must be deleted, i.e. when you delete a Machine Pool, the Machines owned by the MachinePool must be deleted first.

Reply:
I'm not super familiar with the OwnerReference behavior either, so I'm also going off the same docs. However, I'm talking about the opposite relationship, not the deletion of the owners, but the deletion of the dependents.

What I observe and can see of the code is that the AWSMachine is the dependent of the Machine resource. When the AWSMachine is deleted, the owner, the Machine resource remains. It remains until the MachinePool resource that owns the Machine resource is reconciled and detects the Machine as a 'dangling' machine and deletes it.

I'm looking for confirmation if the behavior I described is expected and the behavior desired.

Reply:
Yeah, you got it. The function `deleteDanglingMachines` is meant to clean up Machines that don't correspond to an InfraMachine. Since in this case, we treat the list of InfraMachines as the source of truth of what instances are on the InfraMachinePool.

Reply:
Thanks for the confirmation. Consider this resolved then, as my observed behavior matches expectations here.